daemon/src/db.rs

```diff
@@ -11,17 +13,45 @@ use std::mem;
 use std::path::PathBuf;
 use time::Duration;
 
-pub async fn connect(path: PathBuf) -> Result<SqlitePool> {
-    let pool = SqlitePool::connect_with(
-        SqliteConnectOptions::new()
-            .create_if_missing(true)
-            .filename(path),
-    )
-    .await?;
+/// Connects to the SQLite database at the given path.
+///
+/// If the database does not exist, it will be created. If it does exist, we load it and apply all
+/// pending migrations. If applying migrations fails, the old database is backed up next to it and a
+/// new one is created.
+pub fn connect(path: PathBuf) -> BoxFuture<'static, Result<SqlitePool>> {
+    async move {
+        let pool = SqlitePool::connect_with(
+            SqliteConnectOptions::new()
+                .create_if_missing(true)
+                .filename(&path),
+        )
+        .await?;
 
-    run_migrations(&pool).await?;
+        if let Err(e) = run_migrations(&pool).await {
+            tracing::error!("{:#}", e);
 
-    Ok(pool)
+            let unix_timestamp = time::OffsetDateTime::now_utc().unix_timestamp();
+            let new_path = PathBuf::from(format!("{}-{}-backup", path.display(), unix_timestamp));
+
+            tracing::info!(
+                "Backing up old database at {} to {}",
+                path.display(),
+                new_path.display()
+            );
+
+            tokio::fs::rename(&path, &new_path)
+                .await
+                .context("Failed to rename database file")?;
+
+            tracing::info!("Starting with a new database!");
+
+            // recurse to reconnect (async recursion requires a `BoxFuture`)
+            return connect(path).await;
```

Comment:
🤔 So the assumption is that `run_migrations` cannot fail if the database has just been created, right? Otherwise I can see us creating an endless loop here.

Reply:
Yeah, these are the possible errors: https://docs.rs/sqlx-core/0.5.5/src/sqlx_core/migrate/error.rs.html#3

I guess there are some in here that are not resolved by using a new DB? We should probably only try again on `VersionMismatch` and `VersionMissing`. All others seem permanent!